### DISCUSSION TITLE
Multiple DB support

### DISCUSSION THREAD

**Discussion by crusty-dave** [diesel-rs/diesel#3030] [Q&A] [answered]
I need to support two different PSQL DBs, one of which might end up running in a separate cluster some day.

What is the idiomatic way to support this correctly using diesel?

i.e. I need to ensure that migrations are kept in isolated directories.

I am hoping that I don't have to create a library for one of them, but diesel.toml seems to be restricted to a single DB in a project

TIA, -Dave

**Comment by weiznich** [comment, marked as answer]
> but diesel.toml seems to be restricted to a single DB in a project

That's literally the only place where there is an assumption about a single database connection and that's the only place where it's reasonable to assume such thing. 

Generally speaking: If you use diesel-cli for managing stuff, you just want to use the command line options to set different connections/migration dirs. If you use diesel, it's even easier as different connections can point to different databases. How you actually setup and manage them, that's part of your implementation.

**Reply by crusty-dave** [reply]
FYI: I wasn't worried about diesel, only the CLI and the migrations dir.  I wanted to know the idiomatic way to do this.

What I am going to try is the following:

embed_migrations!("./db/db1/migrations");
embed_migrations!("./db/db2/migrations");

And I am going to try moving diesel.toml to db1/db1.toml & create a new db2/db2.toml and then use the command line arguments. 

```
OPTIONS:
        --config-file <CONFIG_FILE>      The location of the configuration file to use. Falls back to the
                                         `DIESEL_CONFIG_FILE` environment variable if unspecified. Defaults to
                                         `diesel.toml` in your project root. See diesel.rs/guides/configuring-diesel-cli
                                         for documentation on this file.
        --database-url <DATABASE_URL>    Specifies the database URL to connect to. Falls back to the DATABASE_URL
                                         environment variable if unspecified.
```

Thanks

**Reply by crusty-dave** [reply]
@weiznich - it seems that I had some confusion about migrations.

The following documentation is misleading, mod embedded_migrations is actually created by the embed_migrations macro, it would be helpful if the documentation actually stated that to avoid confusion:

https://docs.rs/diesel_migrations/1.4.0/diesel_migrations/macro.embed_migrations.html

Based on the example, I thought I would need to add a use statement that included both the embed_migrations! macro as well as the embedded_migrations module.

TIA, -Dave

**Comment by jacques-kigo** [comment]
I'm a little confused by the answer here. Can you guys please provide an example on how you setup your `diesel.toml` file and `migrations/` folder for two (or more) databases? I also need to connect to two different postgres instances.

From the answer above it looks like you made two different `diesel` files and whenever you want to interact with it you use the cli with the `--config-file` flag to the specific diesel file you want to interact with.

**Reply by crusty-dave** [reply]
I believe that is correct.  I no longer work on that project, so I can't go back and look at any source, but from memory, I believe that my last post about using separate TOML files was the solution.

**Reply by jacques-kigo** [reply]
Some other piecesI was missing was I also needed to specify my nested migrations folder with `--migration-dir <MIGRATION_DIRECTORY>` and database url with `--database-url <DB_URL>`

The migrations_directory is a relative path, If you move your `diesel.toml` file into a folder you need to make sure you update the relative path to your migrations folder.

I ended up making a `diesel/` folder in the root of my project as follows:

```
diesel/db1.diesel.toml
diesel/db2.diesel.toml
diesel/migrations/db1/...
diesel/migrations/db2/...
```
